File: Advance/13.Implement the Tower of Hanoi/Solution.py
def move_disk(src, dest, s, d):
    if not src:
        disk = dest.pop()
        src.append(disk)
        print(f"Move disk {disk} from {d} to {s}")
    elif not dest:
        disk = src.pop()
        dest.append(disk)
        print(f"Move disk {disk} from {s} to {d}")
    elif src[-1] > dest[-1]:
        disk = dest.pop()
        src.append(disk)
        print(f"Move disk {disk} from {d} to {s}")
    else:
        disk = src.pop()
        dest.append(disk)
        print(f"Move disk {disk} from {s} to {d}")


def hanoi(n, src, aux, dest):
    moves = (1 << n) - 1
    s, a, d = "S", "A", "D"

    if n % 2 == 0:
        a, d = d, a

    for i in range(n, 0, -1):
        src.append(i)

    for i in range(1, moves + 1):
        if i % 3 == 1:
            move_disk(src, dest, s, d)
        elif i % 3 == 2:
            move_disk(src, aux, s, a)
        elif i % 3 == 0:
            move_disk(aux, dest, a, d)


if __name__ == "__main__":
    n = int(input("Enter the number of disks: "))
    src, aux, dest = [], [], []
    hanoi(n, src, aux, dest)

File: Advance/13.Implement the Tower of Hanoi/Solution.java
import java.util.Stack;
import java.util.Scanner;

public class TowerOfHanoi {
    private static void moveDisk(Stack<Integer> src, Stack<Integer> dest, char s, char d) {
        int pole1Top = src.isEmpty() ? -1 : src.pop();
        int pole2Top = dest.isEmpty() ? -1 : dest.pop();

        if (pole1Top == -1) {
            src.push(pole2Top);
            System.out.println("Move disk " + pole2Top + " from " + d + " to " + s);
        } else if (pole2Top == -1) {
            dest.push(pole1Top);
            System.out.println("Move disk " + pole1Top + " from " + s + " to " + d);
        } else if (pole1Top > pole2Top) {
            src.push(pole1Top);
            src.push(pole2Top);
            System.out.println("Move disk " + pole2Top + " from " + d + " to " + s);
        } else {
            dest.push(pole2Top);
            dest.push(pole1Top);
            System.out.println("Move disk " + pole1Top + " from " + s + " to " + d);
        }
    }

    private static void hanoi(int n, Stack<Integer> src, Stack<Integer> aux, Stack<Integer> dest) {
        int moves = (1 << n) - 1;
        char s = 'S', d = 'D', a = 'A';

        if (n % 2 == 0) {
            char temp = d;
            d = a;
            a = temp;
        }

        for (int i = n; i >= 1; i--) src.push(i);

        for (int i = 1; i <= moves; i++) {
            if (i % 3 == 1)
                moveDisk(src, dest, s, d);
            else if (i % 3 == 2)
                moveDisk(src, aux, s, a);
            else if (i % 3 == 0)
                moveDisk(aux, dest, a, d);
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter the number of disks: ");
        int n = sc.nextInt();

        Stack<Integer> src = new Stack<>();
        Stack<Integer> aux = new Stack<>();
        Stack<Integer> dest = new Stack<>();

        hanoi(n, src, aux, dest);
    }
}
